src/com/ms/silverking/cloud/dht/daemon/storage/AbstractSegment.java

```diff
@@ -16,111 +16,112 @@
-    /**
-     * Returns whether the operation at the particular offset is an invalidation
-     */
-    boolean isInvalidation(int offset) {
-        if (invalidatedOffsets != null) {
-            return invalidatedOffsets.contains(offset);
-        } else {
-            return MetaDataUtil.isInvalidation(dataBuf, offset + DHTKey.BYTES_PER_KEY);
-        }
-    }
-    
-    long getVersion(int offset) {
-        assert offset >= 0;
-        return MetaDataUtil.getVersion(dataBuf, offset + DHTKey.BYTES_PER_KEY);
-    }
-    
-    long getCreationTime(int offset) {
-        assert offset >= 0;
-        return MetaDataUtil.getCreationTime(dataBuf, offset + DHTKey.BYTES_PER_KEY);
-    }
-    
-    ValueCreator getCreator(int offset) {
-        return MetaDataUtil.getCreator(dataBuf, offset + DHTKey.BYTES_PER_KEY);
+  protected ByteBuffer dataBuf;
+  protected final OffsetListStore offsetListStore;
+  protected final Set<Integer> invalidatedOffsets;
+
+  protected static final int noSuchKey = IntCuckooConstants.noSuchValue;
+  protected static final boolean debugRetrieve = false;
+  protected static final boolean debugPut = false;
+  protected static final boolean debugExternalStore = false;
+
+  AbstractSegment(ByteBuffer dataBuf, OffsetListStore offsetListStore, Set<Integer> invalidatedOffsets) {
+    this.dataBuf = dataBuf;
+    this.offsetListStore = offsetListStore;
+    this.invalidatedOffsets = invalidatedOffsets;
+  }
+
+  abstract int getSegmentNumber();
+
+  protected abstract int getRawOffset(DHTKey key);
+
+  byte[] getChecksum(int offset) {
+    assert offset >= 0;
+    return MetaDataUtil.getChecksum(dataBuf, offset + DHTKey.BYTES_PER_KEY);
+  }
+
+  /**
+   * Returns whether the operation at the particular offset is an invalidation
+   */
+  boolean isInvalidation(int offset) {
+    if (invalidatedOffsets != null) {
+      return invalidatedOffsets.contains(offset);
+    } else {
+      return MetaDataUtil.isInvalidation(dataBuf, offset + DHTKey.BYTES_PER_KEY);
     }
-    
-    int getStoredLength(int offset) {
-        return MetaDataUtil.getStoredLength(dataBuf, offset + DHTKey.BYTES_PER_KEY);
+  }
+
+  long getVersion(int offset) {
+    assert offset >= 0;
+    return MetaDataUtil.getVersion(dataBuf, offset + DHTKey.BYTES_PER_KEY);
+  }
+
+  long getCreationTime(int offset) {
+    assert offset >= 0;
+    return MetaDataUtil.getCreationTime(dataBuf, offset + DHTKey.BYTES_PER_KEY);
+  }
+
+  ValueCreator getCreator(int offset) {
+    return MetaDataUtil.getCreator(dataBuf, offset + DHTKey.BYTES_PER_KEY);
+  }
+
+  int getStoredLength(int offset) {
+    return MetaDataUtil.getStoredLength(dataBuf, offset + DHTKey.BYTES_PER_KEY);
+  }
+
+  private int checkOffset(int offset) {
+    if (offset < 0) {
+      OffsetList offsetList;
+
+      // negative offsets are an index to the list in the offsetListStore
+      try {
+        offsetList = offsetListStore.getOffsetList(-offset);
+      } catch (InvalidOffsetListIndexException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)